library/net.c

```diff
@@ -221,7 +258,41 @@ static int dnet_io_req_queue(struct dnet_net_state *st, struct dnet_io_req *orig
 	int err = 0;
 	struct dnet_io_req *r;
 
-	r = dnet_io_req_copy(st, orig);
+	/*
+	 * If destination is local node there is no need to copy it via net thread.
+	 * Here it could be sheduled to the proper IO pool directly.
+	 * If io_req hold fd instead of data it should be read here.
+	 */
+	if (st == st->n->st) {
+		dnet_log(st->n, DNET_LOG_DEBUG, "Sending to local state: data: 0x%lx data_size: %d", (unsigned long)orig->data, (int)orig->dsize);
+
+		r = dnet_io_req_copy(st, orig, 1);
+		if (!r) {
+			err = -ENOMEM;
+			goto err_out_exit;
+		}
+
+		/*
+		 * Fixup request.
+		 * r->header and r->data are just a pointers to the same buffer.
+		 * Here r->header could be any size, it depend on function that created io_req.
+		 * In net thread r->hsize is ALWAYS sizeof(struct dnet_cmd) and r->data points
+		 * exactly after struct dnet_cmd struct. We need to modify our copy to the same layout
+		 * because dnet_schedule_io requires it.
+		 */
+		if (r->hsize != sizeof(struct dnet_cmd)) {
```

Comment:
I do not see where `dnet_schedule_io` assumes that header is exactly `sizeof(dnet_cmd)` bytes, that would be strange.
It requires pointers to be set correctly, even more, `r->header` pointer at receive side is only needed to print some debug messages. There is no need to manipulate header sizes at all. But they should be set for correctness.

Please drop this check and setup sender code correctly - set `r->header` and `r->data` to point to the same buffer with whole command: `dnet_cmd` + `dnet_exec` + data, and set `r->hsize` to be equal to at least `sizeof struct dnet_cmd`, `r->st` should grab the reference to the receiving state, `r->fd` should be -1.